app/services/order_syncer.rb

```diff
@@ -66,12 +66,16 @@ def update_payment_for(order)
   end
 
   def update_shipment_for(order)
-    shipment = order.shipments.with_state('pending').where(shipping_method_id: shipping_method_id_was).last
+    shipment = pending_shipment_using_shipping_method(order, shipping_method_id_was)
+
     if shipment
-      shipment.update_attributes(shipping_method_id: shipping_method_id)
+      # Find the shipment again to have non-readonly instance.
+      shipment = Spree::Shipment.find(shipment.id)
+
+      select_shipping_method_for_shipment(shipment, shipping_method)
       order.update_attribute(:shipping_method_id, shipping_method_id)
     else
-      unless order.shipments.with_state('pending').where(shipping_method_id: shipping_method_id).any?
+      unless pending_shipment_using_shipping_method(order, shipping_method_id).present?
         order_update_issues.add(order, I18n.t('admin.shipping_method'))
       end
     end
```

Comment:
see order_shipment.rb, in ofn v2 we have one very useful rule (it makes everything much easier): one order can only have one shipment, there will never be 2 shipments in an order. So you can use order.shipment.

also, spree::shipment.selected_shipping_rate is kept to single selection in spree::shipment.selected_shipping_rate_id=
so you can always use order.shipment.selected_shipping_rate

also, you can simply use **order_shipment.shipping_method** that redirects to spree::shipment.shipping_method that redirects to spree::shipment.selected_shipping_rate :-)

I think these facts will help simplify this code, right?